File: frontend/src/api/design.ts
import { z } from "zod";

import { typedInvoke } from "@/lib/tauri";

export const DesignSystemStatusSchema = z.enum([
  "draft",
  "analyzing",
  "schema_ready",
  "ready",
  "updating",
  "failed",
  "archived",
]);

export const DesignSourceRefSchema = z.object({
  project_id: z.string().min(1),
  path: z.string().min(1),
  line: z.number().int().positive().nullable().optional(),
});

export const DesignSystemResponseSchema = z.object({
  id: z.string().min(1),
  primaryProjectId: z.string().min(1),
  name: z.string().min(1),
  description: z.string().nullable(),
  status: DesignSystemStatusSchema,
  currentSchemaVersionId: z.string().nullable(),
  storageRootRef: z.string().min(1),
  createdAt: z.string().min(1),
  updatedAt: z.string().min(1),
  archivedAt: z.string().nullable(),
});

export const DesignSystemSourceResponseSchema = z.object({
  id: z.string().min(1),
  designSystemId: z.string().min(1),
  projectId: z.string().min(1),
  role: z.enum(["primary", "secondary", "reference"]),
  selectedPaths: z.array(z.string()),
  sourceKind: z.enum(["project_checkout", "upload", "url", "manual_note"]),
  gitCommit: z.string().nullable(),
  sourceHashes: z.record(z.string(), z.string()),
  lastAnalyzedAt: z.string().nullable(),
});

export const AgentConversationResponseSchema = z.object({
  id: z.string().min(1),
  context_type: z.string().optional(),
  contextType: z.string().optional(),
  context_id: z.string().optional(),
  contextId: z.string().optional(),
  claude_session_id: z.string().nullable().optional(),
  claudeSessionId: z.string().nullable().optional(),
  provider_session_id: z.string().nullable().optional(),
  providerSessionId: z.string().nullable().optional(),
  provider_harness: z.string().nullable().optional(),
  providerHarness: z.string().nullable().optional(),
  upstream_provider: z.string().nullable().optional(),
  upstreamProvider: z.string().nullable().optional(),
  provider_profile: z.string().nullable().optional(),
  providerProfile: z.string().nullable().optional(),
  title: z.string().nullable(),
  message_count: z.number().optional(),
  messageCount: z.number().optional(),
  last_message_at: z.string().nullable().optional(),
  lastMessageAt: z.string().nullable().optional(),
  created_at: z.string().optional(),
  createdAt: z.string().optional(),
  updated_at: z.string().optional(),
  updatedAt: z.string().optional(),
  archived_at: z.string().nullable().optional(),
  archivedAt: z.string().nullable().optional(),
});

export const AgentMessageResponseSchema = z.object({
  id: z.string().min(1),
  role: z.string().min(1),
  content: z.string(),
  metadata: z.string().nullable(),
  tool_calls: z.unknown().optional(),
  toolCalls: z.unknown().optional(),
  content_blocks: z.unknown().optional(),
  contentBlocks: z.unknown().optional(),
  attribution_source: z.string().nullable().optional(),
  attributionSource: z.string().nullable().optional(),
  provider_harness: z.string().nullable().optional(),
  providerHarness: z.string().nullable().optional(),
  provider_session_id: z.string().nullable().optional(),
  providerSessionId: z.string().nullable().optional(),
  upstream_provider: z.string().nullable().optional(),
  upstreamProvider: z.string().nullable().optional(),
  provider_profile: z.string().nullable().optional(),
  providerProfile: z.string().nullable().optional(),
  logical_model: z.string().nullable().optional(),
  logicalModel: z.string().nullable().optional(),
  effective_model_id: z.string().nullable().optional(),
  effectiveModelId: z.string().nullable().optional(),
  logical_effort: z.string().nullable().optional(),
  logicalEffort: z.string().nullable().optional(),
  effective_effort: z.string().nullable().optional(),
  effectiveEffort: z.string().nullable().optional(),
  input_tokens: z.number().nullable().optional(),
  inputTokens: z.number().nullable().optional(),
  output_tokens: z.number().nullable().optional(),
  outputTokens: z.number().nullable().optional(),
  cache_creation_tokens: z.number().nullable().optional(),
  cacheCreationTokens: z.number().nullable().optional(),
  cache_read_tokens: z.number().nullable().optional(),
  cacheReadTokens: z.number().nullable().optional(),
  estimated_usd: z.number().nullable().optional(),
  estimatedUsd: z.number().nullable().optional(),
  created_at: z.string().optional(),
  createdAt: z.string().optional(),
});

export const DesignSystemDetailResponseSchema = z.object({
  designSystem: DesignSystemResponseSchema,
  sources: z.array(DesignSystemSourceResponseSchema),
  conversation: AgentConversationResponseSchema.nullable(),
});

export const CreateDesignSystemResponseSchema = z.object({
  designSystem: DesignSystemResponseSchema,
  sources: z.array(DesignSystemSourceResponseSchema),
  conversation: AgentConversationResponseSchema,
});

export const DesignStyleguideItemResponseSchema = z.object({
  id: z.string().min(1),
  designSystemId: z.string().min(1),
  schemaVersionId: z.string().min(1),
  itemId: z.string().min(1),
  group: z.enum(["ui_kit", "type", "colors", "spacing", "components", "brand"]),
  label: z.string().min(1),
  summary: z.string(),
  previewArtifactId: z.string().nullable(),
  sourceRefs: z.array(DesignSourceRefSchema),
  confidence: z.enum(["high", "medium", "low"]),
  approvalStatus: z.enum(["needs_review", "approved", "needs_work"]),
  feedbackStatus: z.enum(["none", "open", "in_progress", "resolved", "dismissed"]),
  updatedAt: z.string().min(1),
});

export const DesignStyleguideFeedbackResponseSchema = z.object({
  id: z.string().min(1),
  designSystemId: z.string().min(1),
  schemaVersionId: z.string().min(1),
  itemId: z.string().min(1),
  conversationId: z.string().min(1),
  messageId: z.string().nullable(),
  previewArtifactId: z.string().nullable(),
  sourceRefs: z.array(DesignSourceRefSchema),
  feedback: z.string(),
  status: z.enum(["none", "open", "in_progress", "resolved", "dismissed"]),
  createdAt: z.string().min(1),
  resolvedAt: z.string().nullable(),
});

export const CreateDesignStyleguideFeedbackResponseSchema = z.object({
  feedback: DesignStyleguideFeedbackResponseSchema,
  item: DesignStyleguideItemResponseSchema,
  message: AgentMessageResponseSchema,
});

export const CreateDesignSystemSourceInputSchema = z.object({
  projectId: z.string().min(1),
  role: z.enum(["secondary", "reference"]).optional(),
  selectedPaths: z.array(z.string()).default([]),
});

export const CreateDesignSystemInputSchema = z.object({
  primaryProjectId: z.string().min(1),
  name: z.string().min(1),
  description: z.string().optional(),
  selectedPaths: z.array(z.string()).default([]),
  sources: z.array(CreateDesignSystemSourceInputSchema).default([]),
});

export const CreateDesignStyleguideFeedbackInputSchema = z.object({
  designSystemId: z.string().min(1),
  itemId: z.string().min(1),
  feedback: z.string().min(1),
  conversationId: z.string().optional(),
});

export type DesignSystemResponse = z.infer<typeof DesignSystemResponseSchema>;
export type DesignSystemSourceResponse = z.infer<typeof DesignSystemSourceResponseSchema>;
export type DesignSystemDetailResponse = z.infer<typeof DesignSystemDetailResponseSchema>;
export type CreateDesignSystemInput = z.infer<typeof CreateDesignSystemInputSchema>;
export type CreateDesignSystemResponse = z.infer<typeof CreateDesignSystemResponseSchema>;
export type DesignStyleguideItemResponse = z.infer<typeof DesignStyleguideItemResponseSchema>;
export type DesignStyleguideFeedbackResponse = z.infer<typeof DesignStyleguideFeedbackResponseSchema>;
export type CreateDesignStyleguideFeedbackInput = z.infer<typeof CreateDesignStyleguideFeedbackInputSchema>;
export type CreateDesignStyleguideFeedbackResponse = z.infer<typeof CreateDesignStyleguideFeedbackResponseSchema>;

export const designApi = {
  listProjectDesignSystems: (projectId: string, includeArchived = false) =>
    typedInvoke("list_project_design_systems", { projectId, includeArchived }, z.array(DesignSystemResponseSchema)),

  getDesignSystem: (id: string) =>
    typedInvoke("get_design_system", { id }, DesignSystemDetailResponseSchema.nullable()),

  createDesignSystem: (input: CreateDesignSystemInput) =>
    typedInvoke("create_design_system", { input: CreateDesignSystemInputSchema.parse(input) }, CreateDesignSystemResponseSchema),

  archiveDesignSystem: (id: string) =>
    typedInvoke("archive_design_system", { id }, DesignSystemResponseSchema),

  approveStyleguideItem: (designSystemId: string, itemId: string) =>
    typedInvoke(
      "approve_design_styleguide_item",
      { input: { designSystemId, itemId } },
      DesignStyleguideItemResponseSchema,
    ),

  createStyleguideFeedback: (input: CreateDesignStyleguideFeedbackInput) =>
    typedInvoke(
      "create_design_styleguide_feedback",
      { input: CreateDesignStyleguideFeedbackInputSchema.parse(input) },
      CreateDesignStyleguideFeedbackResponseSchema,
    ),

  resolveStyleguideFeedback: (feedbackId: string) =>
    typedInvoke(
      "resolve_design_styleguide_feedback",
      { input: { feedbackId } },
      DesignStyleguideFeedbackResponseSchema,
    ),
} as const;

File: src-tauri/crates/ralphx-domain/src/entities/design.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

use super::{ChatConversationId, ChatMessageId, ProjectId};

macro_rules! design_id {
    ($name:ident) => {
        #[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
        pub struct $name(pub String);

        impl $name {
            pub fn new() -> Self {
                Self(uuid::Uuid::new_v4().to_string())
            }

            pub fn from_string(s: impl Into<String>) -> Self {
                Self(s.into())
            }

            pub fn as_str(&self) -> &str {
                &self.0
            }
        }

        impl Default for $name {
            fn default() -> Self {
                Self::new()
            }
        }

        impl std::fmt::Display for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}", self.0)
            }
        }
    };
}

design_id!(DesignSystemId);
design_id!(DesignSystemSourceId);
design_id!(DesignSchemaVersionId);
design_id!(DesignStyleguideItemId);
design_id!(DesignStyleguideFeedbackId);
design_id!(DesignRunId);
design_id!(DesignAssetRefId);
design_id!(DesignExportPackageId);

/// Opaque app-owned storage reference, not a raw filesystem path.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct DesignStorageRootRef(pub String);

impl DesignStorageRootRef {
    pub fn from_hash_component(component: impl Into<String>) -> Self {
        Self(component.into())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesignSystemStatus {
    Draft,
    Analyzing,
    SchemaReady,
    Ready,
    Updating,
    Failed,
    Archived,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesignSourceRole {
    Primary,
    Secondary,
    Reference,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesignSourceKind {
    ProjectCheckout,
    Upload,
    Url,
    ManualNote,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesignSchemaVersionStatus {
    Draft,
    Verified,
    Superseded,
    Failed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesignStyleguideGroup {
    UiKit,
    Type,
    Colors,
    Spacing,
    Components,
    Brand,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesignConfidence {
    High,
    Medium,
    Low,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesignApprovalStatus {
    NeedsReview,
    Approved,
    NeedsWork,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesignFeedbackStatus {
    None,
    Open,
    InProgress,
    Resolved,
    Dismissed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesignRunKind {
    Create,
    Update,
    GenerateScreen,
    GenerateComponent,
    ItemFeedback,
    Audit,
    Export,
    Import,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesignRunStatus {
    Queued,
    Running,
    Completed,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DesignSourceRef {
    pub project_id: ProjectId,
    pub path: String,
    pub line: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DesignSystem {
    pub id: DesignSystemId,
    pub primary_project_id: ProjectId,
    pub name: String,
    pub description: Option<String>,
    pub status: DesignSystemStatus,
    pub current_schema_version_id: Option<DesignSchemaVersionId>,
    pub storage_root_ref: DesignStorageRootRef,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub archived_at: Option<DateTime<Utc>>,
}

impl DesignSystem {
    pub fn new(
        primary_project_id: ProjectId,
        name: impl Into<String>,
        storage_root_ref: DesignStorageRootRef,
    ) -> Self {
        let now = Utc::now();
        Self {
            id: DesignSystemId::new(),
            primary_project_id,
            name: name.into(),
            description: None,
            status: DesignSystemStatus::Draft,
            current_schema_version_id: None,
            storage_root_ref,
            created_at: now,
            updated_at: now,
            archived_at: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DesignSystemSource {
    pub id: DesignSystemSourceId,
    pub design_system_id: DesignSystemId,
    pub project_id: ProjectId,
    pub role: DesignSourceRole,
    pub selected_paths: Vec<String>,
    pub source_kind: DesignSourceKind,
    pub git_commit: Option<String>,
    pub source_hashes: BTreeMap<String, String>,
    pub last_analyzed_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DesignSchemaVersion {
    pub id: DesignSchemaVersionId,
    pub design_system_id: DesignSystemId,
    pub version: String,
    pub schema_artifact_id: String,
    pub manifest_artifact_id: String,
    pub styleguide_artifact_id: String,
    pub status: DesignSchemaVersionStatus,
    pub created_by_run_id: Option<DesignRunId>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DesignStyleguideItem {
    pub id: DesignStyleguideItemId,
    pub design_system_id: DesignSystemId,
    pub schema_version_id: DesignSchemaVersionId,
    pub item_id: String,
    pub group: DesignStyleguideGroup,
    pub label: String,
    pub summary: String,
    pub preview_artifact_id: Option<String>,
    pub source_refs: Vec<DesignSourceRef>,
    pub confidence: DesignConfidence,
    pub approval_status: DesignApprovalStatus,
    pub feedback_status: DesignFeedbackStatus,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DesignStyleguideFeedback {
    pub id: DesignStyleguideFeedbackId,
    pub design_system_id: DesignSystemId,
    pub schema_version_id: DesignSchemaVersionId,
    pub item_id: String,
    pub conversation_id: ChatConversationId,
    pub message_id: Option<ChatMessageId>,
    pub preview_artifact_id: Option<String>,
    pub source_refs: Vec<DesignSourceRef>,
    pub feedback: String,
    pub status: DesignFeedbackStatus,
    pub created_at: DateTime<Utc>,
    pub resolved_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DesignRun {
    pub id: DesignRunId,
    pub design_system_id: DesignSystemId,
    pub conversation_id: Option<ChatConversationId>,
    pub kind: DesignRunKind,
    pub status: DesignRunStatus,
    pub input_summary: String,
    pub output_artifact_ids: Vec<String>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub error: Option<String>,
}

impl DesignRun {
    pub fn queued(
        design_system_id: DesignSystemId,
        kind: DesignRunKind,
        input_summary: impl Into<String>,
    ) -> Self {
        Self {
            id: DesignRunId::new(),
            design_system_id,
            conversation_id: None,
            kind,
            status: DesignRunStatus::Queued,
            input_summary: input_summary.into(),
            output_artifact_ids: Vec::new(),
            started_at: None,
            completed_at: None,
            error: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DesignAssetRef {
    pub id: DesignAssetRefId,
    pub design_system_id: DesignSystemId,
    pub source_ref: Option<DesignSourceRef>,
    pub storage_ref: String,
    pub asset_kind: String,
    pub canonical: bool,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DesignExportPackage {
    pub id: DesignExportPackageId,
    pub design_system_id: DesignSystemId,
    pub schema_version_id: Option<DesignSchemaVersionId>,
    pub manifest_artifact_id: String,
    pub package_storage_ref: String,
    pub redacted: bool,
    pub created_at: DateTime<Utc>,
}

#[cfg(test)]
#[path = "design_tests.rs"]
mod tests;
